File: Problems/USACO/chapter1/sec1.2./gift1.py
"""
ID: grgurin1
LANG: PYTHON3
TASK: gift1
"""


with open('gift1.in', 'r') as fin:
    friends = {}
    np = int(fin.readline().strip())
    for _ in range(np):
        friends[fin.readline().strip()] = 0
    
    for _ in range(np):
        giver = fin.readline().strip()
        amount, recive = map(int, fin.readline().strip().split())

        # Čisti data types, nema puno algoa
        if recive != 0:
            if amount != 0:
                friends[giver] += -amount + (amount%recive)
                amount = amount//recive

            for _ in range(recive):
                friends[fin.readline().strip()] += amount

with open('gift1.out', 'w') as fout:
    for fri, amo in friends.items():
        fout.write(f'{fri} {amo}' + '\n')

File: Problems/USACO/chapter1/sec1.2./gift1.cpp
/*
ID: grgurin1
TASK: gift1
LANG: C++                 
*/
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>

using namespace std;
unordered_map<string, int> givers;
int n;

int main() {
    ifstream fin("gift1.in");
    ofstream fout("gift1.out");

    fin >> n;
    string people[n];
    for(int i = 0; i<n; i++) {
        string name;
        fin >> name;
        people[i] = name;
        givers[name] = 0;
    }

    for(int i = 0; i<n; i++) {
        string giver;
        fin >> giver;
        
        int amount, times;
        fin >> amount >> times;
        if(times) {
            givers[giver] -= (amount - amount%times);
            amount = (int)amount / times;

            for(int j = 0; j<times; j++) {
                string name;
                fin >> name;
                givers[name] += amount;
            }
        }

    }

    for(auto item : people) {
        fout << item << " " << givers[item] << "\n";
    }
}